### PR TITLE
Add slides on virtual destructors

### PR DESCRIPTION
**Pull request by bernhardmgruber**
Fixes: #99

Also includes a small rewording of the slide on overriding.

### CODE REVIEW

File: talk/objectorientation/advancedoo.tex [hunk 1]
```diff
@@ -121,8 +121,9 @@
   \begin{block}{the concept}
     \begin{itemize}
     \item methods can be declared \mintinline{cpp}{virtual}
-    \item for these, the most derived object is always considered
-    \item for others, the type of the variable decides
+    \item for these, the most derived object's implementation is used.
+          (i.e. the dynamic type behind a pointer/reference)
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:125]:
```suggestion
          (i.e.\ the dynamic type behind a pointer/reference)
```
😄

File: talk/objectorientation/advancedoo.tex [hunk 2]
```diff
@@ -121,8 +121,9 @@
   \begin{block}{the concept}
     \begin{itemize}
     \item methods can be declared \mintinline{cpp}{virtual}
-    \item for these, the most derived object is always considered
-    \item for others, the type of the variable decides
+    \item for these, the most derived object's implementation is used.
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:124]:
```suggestion
    \item for these, the most derived object's implementation is used
```

File: talk/objectorientation/advancedoo.tex [hunk 3]
```diff
@@ -305,6 +306,54 @@
...
+      struct Drawable {
+         virtual void draw() = 0;
+      };
+
+      Drawable* p = getImpl(); // returns derived obj.
+      p->draw();
+      delete p;
+      // better interface (discussed later):
+      // std::unique_ptr<Drawable> p = getImpl();
+      // p->draw();
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:324]:
```suggestion
```

Don't do this here. It's too much to read, and we're not there yet.
If you want to keep it, make it a thing that's being uncovered on the next slide -- à la:
BTW, this is how you do it today [...]

But even that might not be such a good style. 🤔

File: talk/objectorientation/advancedoo.tex [hunk 4]
```diff
@@ -305,6 +306,54 @@
...
+    \end{cppcode*}
+    \item What happens when \mintinline{cpp}{p} is deleted?
+    \item What if the class deriving from \mintinline{cpp}{Drawable} has a destructor?
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:327]:
```suggestion
    \item What if a class deriving from \mintinline{cpp}{Drawable} has a destructor?
```

File: talk/objectorientation/advancedoo.tex [hunk 5]
```diff
@@ -305,6 +306,54 @@
...
+
+      Interface* p = getImpl(); // returns derived obj.
+      p->draw();
+      delete p; // dynamic dispatch to right destructor
+    \end{cppcode*}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Best practice}
+    \begin{itemize}
+      \item If you expect users to inherit from your class and override methods (i.e. use your class polymorphically), declare its destructor \mintinline{cpp}{virtual}
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:352]:
```suggestion
      \item If you expect users to inherit from your class and override methods (i.e.\ use your class polymorphically), declare its destructor \mintinline{cpp}{virtual}
```

File: talk/objectorientation/advancedoo.tex [hunk 6]
```diff
@@ -305,6 +306,54 @@
...
+    \begin{itemize}
+    \item We sometimes need to maintain owning pointers to base classes:
+    \begin{cppcode*}{gobble=2}
+      struct Drawable {
+         virtual void draw() = 0;
+      };
+
+      Drawable* p = getImpl(); // returns derived obj.
+      p->draw();
+      delete p;
```
Comment (hageboeck) [inline talk/objectorientation/advancedoo.tex:321]: Maybe put all the code into an `exampleblock`? I don't really insist, but I think most of the time we have the blue boxes for text and another block for code.

File: talk/objectorientation/advancedoo.tex (hunk 3 above)
Comment (hageboeck) [reply talk/objectorientation/advancedoo.tex:324]: Referring to the exampleblock comment above, you could just show the new exampleblock with the new code if you want to keep the comment about the unique_ptr.

**Review by hageboeck — COMMENTED** [review]
Mostly small comments, but let's reconsider mentioning the unique_ptr here.

File: talk/objectorientation/advancedoo.tex (hunk 6 above)
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex:321]: I should at least put it out of the block of the bullet point list.

File: talk/objectorientation/advancedoo.tex (hunk 3 above)
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex]: I put it onto a separate slide using `overprint`. What do you think?

File: talk/objectorientation/advancedoo.tex (hunk 6 above)
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex]: Better?

File: talk/objectorientation/advancedoo.tex (hunk 3 above)
Comment (hageboeck) [reply talk/objectorientation/advancedoo.tex]: Better! 👍

File: talk/objectorientation/advancedoo.tex (hunk 6 above)
Comment (hageboeck) [reply talk/objectorientation/advancedoo.tex]: Yes!

**Review by hageboeck — APPROVED** [review]
LGTM

File: talk/objectorientation/advancedoo.tex (hunk 3 above)
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex]: Great!

**Pull request merged**